application/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+            tokio::select! {
+                ctrl_c_signal = tokio::signal::ctrl_c() => {
+                    if ctrl_c_signal.is_err() {
+                        log::info!("Ctrl-C signal handler failed");
+                    }
+                }
+                () = shutdown_rq.recv() => {},
+                call = call_rq.recv() => { call(&mut ()); }
+            };
+        });
+
+        log::info!("Application {} running", self.name);
+
+        // Wait for a subsystem to shut down.
+        self.shutting_down_rx.recv().await.expect("All subsystems already dead");
+
+        // Drop the unused shutdown request end
+        std::mem::drop(self.shutdown_request_rx);
+        // Signal the application is shut down so it does not wait for itself
+        std::mem::drop(self.shutting_down_tx);
+
+        log::info!("Application {} shutting down", self.name);
+
+        // Order all the remaining subsystems to shut down.
+        let _ = self.shutdown_request_tx.send(());
+
+        // Wait for the subsystems to go down.
+        // TODO add shutdown timeout here
+        Self::wait_for_subsystems_to_shut_down(self.shutting_down_rx).await;
+
+        log::info!("Application {} terminated", self.name);
+    }
+}
+
+/// Internal action type sent in the channel.
+type Action<T> = Box<dyn FnOnce(&mut T) + Send>;
+
+/// Call request
+pub struct CallRequest<T>(mpsc::Receiver<Action<T>>);
+
+impl<T: 'static + Send> CallRequest<T> {
+    /// Receive an external call to this subsystem.
+    pub async fn recv(&mut self) -> Action<T> {
+        match self.0.recv().await {
+            // We have a call, return it
+            Some(action) => action,
+            // All handles to this subsystem dropped, suspend call handling.
+            None => std::future::pending().await,
+        }
+    }
+}
+
+/// Shutdown request
+pub struct ShutdownRequest(broadcast::Receiver<()>);
+
+impl ShutdownRequest {
+    /// Receive a shutdown request.
+    pub async fn recv(&mut self) {
+        match self.0.recv().await {
+            Err(broadcast::error::RecvError::Lagged(_)) => {
+                panic!("Multiple shutdown broadcast requests issued")
+            }
+            Err(broadcast::error::RecvError::Closed) => {
+                panic!("Shutdown channel sender closed prematurely")
+            }
+            Ok(()) => (),
+        }
+    }
+}
+
+/// Subsystem handle.
+///
+/// This allows the user to interact with the subsystem from the outside. Currently, it only
+/// supports calling functions on the subsystem.
+#[derive(Clone)]
+pub struct Subsystem<T> {
+    // Send the subsystem stuff to do.
+    action_tx: mpsc::Sender<Action<T>>,
+}
+
+impl<T: Send + 'static> Subsystem<T> {
+    /// Crate a new subsystem handle.
+    fn new(action_tx: mpsc::Sender<Action<T>>) -> Self {
+        Self { action_tx }
+    }
+
+    /// Dispatch a function call to the subsystem
+    pub async fn call_mut<R: Send + 'static>(
```

Comment:
Is it possible to create non-blocking interface so, e.g., P2P could submit a bunch of blocks or maybe a block with PP so it takes non-trivial amount of computation (= time) to validate it and it wouldn't have to block itself until the validation is done but instead would receive like a receipt that the blocks were processed correctly/they were invalid and the connection should be closed?

Reply:
Maybe there's something I don't understand. But doesn't the `async` keyword do exactly that by returning a future of the result, acting like a receipt that you can poll later?

Reply:
Unless I'm reading the code incorrectly, that's not really what happens here. The code is asynchronous on "executor-level" but not on "application-level" meaning while my task is able to sleep and let somebody else run because I'm polling a future, I cannot progress my own state because I'm polling the future, if that makes sense. This is because calling the function (`mpsc::Sender::send()`) and fetching the return value (`oneshot::Receiver:recv()`) happen inside this function so I can't separate the "making a request" and "getting a response" in my application logic. If it would instead return the oneshot receiver, then at least I could decided myself when I want that to poll the future that gives me the return value but as you probably can imagine, it becomes really messy with multiple futures.

If there was instead an API like `submit_work(work_id: u64, work: vec![..])` and I would have an RX channel that would poll for `WorkResponse { work_id: u64, status: WorkStatus }`, then I could asynchronously submit work to mempool and chainstate and get the response back when they are ready, without having to poll the return value future in my application logic.

This is at least how I understood the current code.

Reply:
Maybe I'm also ignorant enough to be saying nonsense. But here's what I understand:

When we call this function, it immediately returns and gives a future object. 

> I cannot progress my own state because I'm polling the future, if that makes sense

Once you have the future object, polling it isn't a blocking operation unless you call "await". You can keep asking it whenever you want "hey, buddy, are you ready?", and if not, you can do whatever other things you want to do. Once it's ready, it can be either an error or a result. At least this is how I did futures before. I'm failing to understand the problem, but maybe you know this better than me.

Reply:
> I can change this to return the oneshot receiver to give the caller more flexibility if desired.

While I don't understand the real problem still, this seems to be the wrong abstraction for this problem.

Reply:
I think we should have a call about this problem. I'm not happy about exposing the internals of a subsystem to the contents of other subsystems, or maybe that's not even what's being proposed here. For example, we should totally be able to change the channels' types to another kind of channels without having to touch any users of the subsystems. Let's have that call at some time and maybe you guys can make me understand the issue better.

Reply:
The idea was to make the approach easiest for the user (i.e. one that is the closest to the regular function calls) the default one. The P2P system is somewhat different as it is highly concurrent already. There is not an issue having both in the API here as far as I'm concerned.

Reply:
The conclusion from the call was to keep the interface to the `call` method as it is now. It's a responsibility of the client to spawn tasks appropriately if more concurrency is required. If that becomes too burdensome, we can expose the channel-based API too at some point.